### ISSUE TITLE
Color spaces are not tree-shakeable

### ISSUE THREAD

**Issue by LeaVerou** [LeaVerou/color.js#163]
(Remaining part of #159)

While the new procedural API is tree-shakeable wrt functions, since color spaces register themselves onto `ColorSpace.registry`, this means they are not tree-shakeable.

To make them tree-shakeable, we'd need to eliminate side effects from the color space modules. Then, the importing code would register them.

Something like:

```js
import ColorSpace from "COLORJS_ROOT/src/space.js";
import oklch from "COLORJS_ROOT/src/spaces/oklch.js";
ColorSpace.register(oklch);
```

We could make this less painful by offering pre-generated indices that do this for all existing color spaces, but I'm still struggling with the idea of requiring more plumbing to use Color.js, which makes it easier to make mistakes. E.g. someone could easily forget to register their color space and do something like:

```js
import oklch from "COLORJS_ROOT/src/spaces/oklch.js";
let color = new Color("oklch", [.5, .2, 340]); // TypeError: No color space found with id = "oklch"
```

Note that registration is needed only to use string ids instead of color space objects. The API is perfectly usable with just passing identifiers around and never registering any color space, like so:

```js
import srgb from "COLORJS_ROOT/src/spaces/srgb.js";
import oklch from "COLORJS_ROOT/src/spaces/oklch.js";
let color = new Color(oklch, [.5, .2, 340]);
color.to(srgb);
```

Thoughts?

**Comment by sgomes** [comment]
Thank you for spending so much time working on a procedural API 🙇 That massively helps with keeping bundle sizes down!

> Note that registration is needed only to use string ids instead of color space objects.

Yes, that's what I see as the greatest difficulty in using the current color space approach for everything. Would it be feasible to make parsing and serialization self-sufficient within each colour space?

The way the top-level `parse` and `serialize` functions are generated is already really modular, so it may be a case of making a couple of utils that are just a bit more generic than that, and using them in every colour space to generate standalone versions there?

**Comment by sgomes** [comment]
> Perhaps it could accept an optional list of color spaces that will be used instead of ColorSpace.all / ColorSpace.registry?

Yeah, an optional list of color spaces would likely do the trick here 👍 That would allow for use of the function while having an empty registry.

`serialize` doesn't need it, as you mention, but perhaps it could do the same, for symmetry? Though I'm really not sure. I can't decide whether it would indeed make things easier to use or just unnecessarily introduce more complexity 🙂

**Comment by lydell** [comment]
FYI https://colorjs.io/docs/spaces.html says that this issue is open – should that be removed from the page?

<img width="1051" alt="image" src="https://user-images.githubusercontent.com/2142817/184670286-3d202907-f867-4846-9fa4-a3083e81e154.png">

**Comment by LeaVerou** [comment]
Oops, yes! PR?

**Comment by lydell** [comment]
> PR?

I’m sorry, but I’ll pass on that this time! 😊

**Comment by LeaVerou** [comment]
Can you open a new issue then since this is closed?

**Comment by lydell** [comment]
Sure thing! Opened a new issue: https://github.com/LeaVerou/color.js/issues/204
